README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# genderBR

[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/genderBR)](https://cran.r-project.org/package=genderBR)
[![Travis-CI Build
Status](https://travis-ci.org/meirelesff/genderBR.svg?branch=master)](https://travis-ci.org/meirelesff/genderBR)
[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/meirelesff/genderBR?branch=master&svg=true)](https://ci.appveyor.com/project/meirelesff/genderBR)
[![Package-License](https://img.shields.io/badge/License-GPL-brightgreen.svg)](http://www.gnu.org/licenses/gpl-2.0.html)

`genderBR` predicts gender from Brazilian first names using data from
the Instituto Brasileiro de Geografia e Estatistica’s 2010 Census.

## How does it work?

`genderBR`’s main function is `get_gender`, which takes a string with a
Brazilian first name and predicts its gender using data from the IBGE’s
2010 Census – specifically, from its API and from an internal dataset.

More specifically, it uses data on the number of females and males with
the same name in Brazil, or in a given Brazilian state, and calculates
the proportion of female’s uses of it. The function then classifies a
name as male or female only when that proportion is higher than a given
threshold (e.g., `female if proportion > 0.9`, or `male if proportion
<= 0.1`); proportions below those threshold are classified as missing
(`NA`). An example:

``` r
library(genderBR)
#> 
#> To cite genderBR in publications, use: citation('genderBR')
#> To learn more, visit: http://fmeireles.com/genderbr

get_gender("joão")
#> [1] "Male"
get_gender("ana")
#> [1] "Female"
```

Multiple names can be passed at the same function call:

``` r
get_gender(c("pedro", "maria"))
#> [1] "Male"   "Female"
```

And both full names and names written in lower or upper case are
accepted as inputs:

``` r
get_gender("Mario da Silva")
#> [1] "Male"
get_gender("ANA MARIA")
#> [1] "Female"
```

Additionally, one can filter results by state with the argument `state`;
or get the probability that a given first name belongs to a female
person by setting the `prob` argument to `TRUE` (defaults to `FALSE`).

``` r
# What is the probability that the name Ariel belongs to a female person in Brazil?
get_gender("Ariel", prob = TRUE)
#> [1] 0.09219289

# What about differences between Brazilian states?
get_gender("Ariel", prob = TRUE, state = "RJ") # RJ, Rio de Janeiro
#> [1] 0.2627399
get_gender("Ariel", prob = TRUE, state = "RS") # RS, Rio Grande do Sul
#> [1] 0.05144695
get_gender("Ariel", prob = TRUE, state = "SP") # SP, Sao Paulo
#> [1] 0.1294782
```

Note that a vector with states’ abbreviations is a valid input for
`get_gender` function, so this also works:

``` r
name <- rep("Ariel", 3)
states <- c("rj", "rs", "sp")
get_gender(name, prob = T, state = states)
#> [1] 0.26273991 0.05144695 0.12947819
```

This can be useful also to predict the gender of different individuals
living in different states:

``` r
df <- data.frame(name = c("Alberto da Silva", "Maria dos Santos", "Thiago Rocha", "Paula Camargo"),
                 uf = c("AC", "SP", "PE", "RS"),
                 stringsAsFactors = FALSE
                 )

df$gender <- get_gender(df$name, df$uf)

df
#>               name uf gender
#> 1 Alberto da Silva AC   Male
#> 2 Maria dos Santos SP Female
#> 3     Thiago Rocha PE   Male
#> 4    Paula Camargo RS Female
```

### Brazilian state abbreviations

The `genderBR` package relies on Brazilian state abbreviations
(acronyms) to filter results. To get a complete dataset with the full
name, IBGE code, and abbreviations of all 27 Brazilian states, use the
`get_states` functions:

``` r
get_states()
#> # A tibble: 27 x 3
#>    state            abb    code
#>    <chr>            <chr> <int>
#>  1 ACRE             AC       12
#>  2 ALAGOAS          AL       27
#>  3 AMAPA            AP       16
#>  4 AMAZONAS         AM       13
#>  5 BAHIA            BA       29
#>  6 CEARA            CE       23
#>  7 DISTRITO FEDERAL DF       53
#>  8 ESPIRITO SANTO   ES       32
#>  9 GOIAS            GO       52
#> 10 MARANHAO         MA       21
#> # ... with 17 more rows
```

## Geographic distribution of Brazilian first names

The `genderBR` package can also be used to get information on the
relative and total number of persons with a given name by gender and by
state in Brazil. To that end, use the `map_gender` function:

``` r
map_gender("maria")
#> Warning: `as.tbl()` is deprecated as of dplyr 1.0.0.
#> Please use `tibble::as_tibble()` instead.
#> This warning is displayed once every 8 hours.
#> Call `lifecycle::last_warnings()` to see where this warning was generated.
#> # A tibble: 27 x 6
#>    nome                   uf    freq populacao sexo    prop
#>    <chr>               <int>   <int>     <int> <chr>  <dbl>
#>  1 Piauí                  22  363139   3118360 ""    11645.
#>  2 Ceará                  23  967042   8452381 ""    11441.
#>  3 Paraíba                25  423026   3766528 ""    11231.
#>  4 Rio Grande do Norte    24  341940   3168027 ""    10793.
#>  5 Alagoas                27  321330   3120494 ""    10297.
#>  6 Pernambuco             26  838534   8796448 ""     9533.
#>  7 Sergipe                28  188619   2068017 ""     9121.
#>  8 Maranhão               21  574689   6574789 ""     8741.
#>  9 Acre                   12   63172    733559 ""     8612.
#> 10 Minas Gerais           31 1307650  19597330 ""     6673.
#> # ... with 17 more rows
```

To specify gender in the consultation, use the optional argument
`gender` (valid inputs are `f`, for female; `m`, for male; or `NULL`,
the default option).

``` r
map_gender("iris", gender = "m")
#> No encoding supplied: defaulting to UTF-8.
#> # A tibble: 23 x 6
#>    nome                uf  freq populacao sexo   prop
#>    <chr>            <int> <int>     <int> <chr> <dbl>
#>  1 Goiás               52   840   6003788 m     14.0 
#>  2 Tocantins           17   156   1383445 m     11.3 
#>  3 Bahia               29   422  14016906 m      3.01
#>  4 Mato Grosso         51    91   3035122 m      3   
#>  5 Minas Gerais        31   512  19597330 m      2.61
#>  6 Distrito Federal    53    65   2570160 m      2.53
#>  7 Espírito Santo      32    69   3514952 m      1.96
#>  8 Rondônia            11    28   1562409 m      1.79
#>  9 Pará                15   129   7581051 m      1.7 
#> 10 Rio de Janeiro      33   225  15989929 m      1.41
#> # ... with 13 more rows
```

## Installing

To install `genderBR`’s last stable version on CRAN, use:

``` r
install.packages("genderBR")
```

To install a development version, use:

``` r
if (!require("devtools")) install.packages("devtools")
devtools::install_github("meirelesff/genderBR")
```

## Data

The surveyed population in the Instituto Brasileiro de Geografia e
Estatistica’s (IBGE) 2010 Census includes 190,8 million Brazilians –
with more than 130,000 unique first names.

To extracts the numer of male or female uses of a given first name in
Brazil, the package employs the IBGE’s
[API](https://censo2010.ibge.gov.br/nomes/) and, from in 1.1.0 version,
also from an internal dataset containing all the names recorded in the
IBGE’s Census. In this service, different spelling (e.g., Ana and Anna,
or Marcos and Markos) implies different occurrences, and only names with
more than 20 occurrences, or more than 15 occurrences in a given state,
are included in the database.

For more information on the IBGE’s data, please check (in Portuguese):
<https://censo2010.ibge.gov.br/nomes/>

## Author

[Fernando Meireles](https://fmeireles.com)


NEWS.md
# genderBR 1.1.1

In this version, a few improvements and bug fixed were introduced. Most important, connection errors now return informative messages to users.

- `map_gender` and `get_gender` now return informative error messages when reach timeout
- `get_gender` function better handles non-ASCII characters
- Documentation expanded to notify users that IBGE's API does not work with UTF-8 special characters
- Magritte's pipe exported


# genderBR 1.1.0

In this minor release, the genderBR package was improved in two ways. First, bugs and some minor issues were fixed, making the package's functions more stable. Second, the package now contains an internal dataset with all the names reported by the IBGE's Census that is used by the get_gender function to predict gender from Brazilian first names. Therefore, classifying a vector with more than 1,000 names takes no more than a few seconds now. Overall, these are the improvements:

* Added a `NEWS.md` file to track changes to the package.
* Added input checks to the `get_gender` function.
* Reduce the time between requests to the IBGE's Census API.
* Fixed a problem on vectorization in the internal `round_guess` funcion.
* Included an internal dataset with all Brazilian first names and their predicted gender extracted from the IBGE.
* Update the `get_gender` function to work with internal data.





R/utils.R
# OnAttach message
.onAttach <-
  function(libname, pkgname) {
    packageStartupMessage("\nTo cite genderBR in publications, use: citation('genderBR')")
    packageStartupMessage("To learn more, visit: fmeireles.com/genderbr\n")
  }


#' State's abbreviations
#'
#' Use this function to get a \code{data.frame} with the full names, abbreviations
#' (acronym), and IBGE codes of all Brazilian states.
#'
#' @return A \code{tbl_df, tbl, data.frame} with two variables: \code{state}, \code{abb}, and \code{code}.
#' @export

get_states <- function(){

  tibble::tibble(state = c("ACRE", "ALAGOAS", "AMAPA", "AMAZONAS", "BAHIA", "CEARA", "DISTRITO FEDERAL",
                       "ESPIRITO SANTO", "GOIAS", "MARANHAO", "MATO GROSSO DO SUL", "MATO GROSSO",
                       "MINAS GERAIS", "PARA", "PARAIBA", "PARANA", "PERNAMBUCO", "PIAUI",
                       "RIO DE JANEIRO", "RIO GRANDE DO NORTE", "RIO GRANDE DO SUL", "RONDONIA",
                       "RORAIMA", "SANTA CATARINA", "SAO PAULO", "SERGIPE", "TOCANTINS"),
             abb = c("AC", "AL", "AP", "AM", "BA", "CE", "DF", "ES", "GO", "MA",
                     "MS", "MT", "MG", "PA", "PB", "PR","PE", "PI", "RJ", "RN",
                     "RS", "RO", "RR", "SC", "SP", "SE", "TO"),
             code = c(12L, 27L, 16L, 13L, 29L, 23L, 53L, 32L, 52L, 21L, 51L, 50L,
                      31L, 15L, 25L, 41L, 26L, 22L, 33L, 24L, 43L, 11L, 14L, 42L, 35L,
                      28L, 17L))
}


# Internal function to test void names
test_responses <- function(response1, response2, prob){

  httr::stop_for_status(response1)
  httr::stop_for_status(response2)

  if(length(response1$content) == 2 & length(response2$content) == 2) return(NA)
  if(length(response1$content) == 2 & length(response2$content) > 2 & prob == TRUE) return(0)
  if(length(response1$content) > 2 & length(response2$content) == 2 & prob == TRUE) return(1)
  if(length(response1$content) == 2 & length(response2$content) > 2 & prob == FALSE) return("Male")
  if(length(response1$content) > 2 & length(response2$content) == 2 & prob == FALSE) return("Female")

  NULL
}


# Internal function to round numeric guess
round_guess <- function(prob, threshold){

  dplyr::case_when(prob > threshold ~ "Female",
                   prob < (1 - threshold) ~ "Male",
                   TRUE ~ as.character(NA)
                   )
}


# Internal function to clean first names
clean_names <- function(name){

  sub("(.*?) .*", "\\1", name) %>%
    iconv(to = "ASCII//TRANSLIT") %>%
    tolower()
}


# Internal function to prepare vector with state abbreviations
get_state <- function(state, ln){

  state <- sapply(state, function(state) state2code(state))
  if(ln > 1 & length(state) == 1) state <- rep(state, ln)
  state
}


# Internal function to match state abbreviations and state codes
state2code <- function(uf){

  ufs <- get_states()$abb

  uf <- toupper(uf) %>%
    match.arg(ufs)

  get_states()$code[match(uf, ufs)]
}


# Safe GET (to avoid unninformative timeouts)
get_safe <- purrr::possibly(httr::GET, otherwise = NULL)






R/map_gender.R
#' Map the use of Brazilian first names by gender and by state
#'
#' \code{map_gender} retrieves data on the number of male or female uses of a given first name
#' by state from the Instituto Brasileiro de Geografia e Estatistica's 2010 Census API.
#'
#' @param name A string with a Brazilian first name. The name can also be passed to the function
#' as a full name (e.g., Ana Maria de Souza). \code{get_gender} is case insensitive.
#' @param gender A string with the gender to look for. Valid inputs are \code{m}, for males, \code{f},
#' for females, and \code{NULL}, in which case the function returns results for all persons with a given name.
#'
#' @details Information on the gender associated with Brazilian first names was collect in the 2010 Census
#' (Censo Demografico de 2010, in Portuguese), in July of that year, by the Instituto Brasileiro de Demografia
#' e Estatistica (IBGE). The surveyed population includes 190,8 million Brazilians living in all 27 states.
#' According to the IBGE, there are more than 130,000 unique first names in this population.
#'
#' @note Names with different spell (e.g., Ana and Anna, or Marcos and Markos) are considered different names.
#' Additionally, only names with more than 20 occurrences, or more than 15 occurrences in a given state,
#' are considered.
#'
#' @references For more information on the IBGE's data, please check (in Portuguese):
#' \url{https://censo2010.ibge.gov.br/nomes/}
#'
#' @seealso \code{\link{get_gender}}
#'
#' @return \code{get_gender} returns a \code{tbl_df, tbl, data.frame} with the following variables:
#'
#' \itemize{
#'   \item \code{nome} State's name.
#'   \item \code{uf} State's abbreviation.
#'   \item \code{freq} Total number of persons with the name provided.
#'   \item \code{populacao} State's total population.
#'   \item \code{sexo} Same as the \code{sexo} argument provided.
#'   \item \code{prop} Persons with the name and gender provided per 100,000 inhabitants.
#' }
#'
#' @examples
#' # Map the use of the name 'Joao' for
#' # males in Brazil by state
#' map_gender('Joao', gender = 'm')
#'
#' \donttest{
#' # Map the use of the name 'Maria'
#' map_gender('maria')
#'
#' # The function accepts full names
#' map_gender('Maria da Silva Santos')
#'
#' # Or names in uppercase
#' map_gender('MARIA DA SILVA SANTOS')
#' }
#'
#' \dontshow{
#' map_gender('AUGUSTO ROBERTO', gender = 'm')
#' map_gender('John da Silva', gender = 'm')
#' }
#'
#'
#' @export


map_gender <- function(name, gender = NULL){


  # Clean name
  name <- clean_names(name)

  # GET
  total <- "https://servicodados.ibge.gov.br/api/v1/censos/nomes/mapa" %>%
    get_safe(query = list(nome = name, sexo = gender))

  # Test response
  if(is.null(total)) stop("IBGE's API is not responding. Try again later.")
  httr::stop_for_status(total)

  # Parse and return
  httr::content(total, as = "text") %>%
    jsonlite::fromJSON() %>%
    tibble::as_tibble()
}


R/get_gender.R
#' Predict gender from Brazilian first names
#'
#' @description
#' \code{get_gender} uses the IBGE's 2010 Census data to predict gender from Brazilian first names.
#' More specifically, it uses data on the number of females and males with the same name
#' in Brazil, or in a given Brazilian state, and calculates the proportion of females using it.
#' The function classifies a name as male or female only when that proportion is higher than
#' a given threshold (e.g., \code{female if proportion > 0.9, the default}, or \code{male if proportion < 0.1});
#' proportions below this threshold are classified as missing (\code{NA}). This method is based on the 'gender'
#' package developed by Muellen (2016). gender: Predict Gender from Names Using Historical Data.
#'
#' Multiple names can be passed to the function call. To speed the calculation process,
#' the package aggregates equal first names to make fewer requests to the IBGE's API. Also, the package contains an internal dataset with all the names reported by the IBGE to make faster classifications -- although this option does not support getting results by State.
#'
#' @param names A string specifying a person's first name. Names can also be passed to the function
#' as a full name (e.g., Ana Maria de Souza). \code{get_gender} is case insensitive.
#' @param state A string with the state of federation abbreviation (e.g., \code{RJ} for Rio de Janeiro). If state is set to a value different from \code{NULL}, the \code{internal} argument is ignored.
#' @param prob Report the proportion of female uses of the name? Defaults to \code{FALSE}.
#' @param threshold Numeric indicating the threshold used in predictions. Defaults to 0.9.
#' @param internal Use internal data to predict gender? Allowing this option makes the function way faster, but it does not support getting results by State. Defaults to \code{TRUE}.
#'
#' @details Information on the Brazilian first names uses by gender was collect in the 2010 Census
#' (Censo Demografico de 2010, in Portuguese), in July of that year, by the Instituto Brasileiro de Demografia
#' e Estatistica (IBGE). The surveyed population includes 190,8 million Brazilians living in all 27 states.
#' According to the IBGE, there are more than 130,000 unique first names in this population.
#'
#' @note Names with different spell (e.g., Ana and Anna, or Marcos and Markos) are considered different names.
#' In addition, only names with more than 20 occurrences, or more than 15 occurrences in a given state,
#' are included in the IBGE's data.
#'
#' Also note that UTF-8 special characters, common in Portuguese words and names, are not supported by the IBGE's API.
#' Users are encouraged to convert strings to ASCII.
#'
#' @references For more information on the IBGE's data, please check (in Portuguese):
#' \url{https://censo2010.ibge.gov.br/nomes/}
#'
#' @seealso \code{\link{map_gender}}
#'
#' @return \code{get_gender} may returns three different values: \code{Female}, if the name provided is female;
#' \code{Male}, if the name provided is male; or \code{NA}, if we can not predict gender from the name given the chosen threshold.
#'
#' If the \code{prob} argument is set to \code{TRUE}, then the function returns the proportion of females uses of the provided name.
#'
#' @examples
#' # Use get_gender to predict the gender of a person based on her/his first name
#' get_gender('MARIA DA SILVA SANTOS')
#' get_gender('joao')
#'
#' \dontrun{
#' # It is possible to filter results by state
#' get_gender('ana', state = 'sp')
#'
#' # To change the employed threshold
#' get_gender('ariel', threshold = '0.8')
#'
#' # Or to get the proportion of females
#' # with the name provided
#' get_gender('iris', prob = TRUE)
#'
#' # Multiple names can be predict at the same time
#' get_gender(c('joao', 'ana', 'benedita', 'rafael'))
#'
#' # In different states
#' get(rep('cris', 3), c('sp', 'am', 'rs'))
#' }
#'
#' @export

get_gender <- function(names, state = NULL, prob = FALSE, threshold = 0.9, internal = TRUE){


  # Inputs
  if(!prob & threshold < 0 | !prob & threshold > 1) stop("Threshold must be between 0 and 1.")
  if(!is.logical(internal)) stop("Internal must be logical.")
  if(!is.logical(prob)) stop("Prob must be logical.")

  # Names
  names <- clean_names(names)
  un_names <- unique(names)
  ln_un <- length(un_names)
  ln <- length(names)

  # Set pauses
  if(ln > 100) pause <- TRUE
  else pause <- FALSE


  # Ignore internal when state is declared
  if(internal & !is.null(state)) internal <- FALSE


  ### Internal data
  if(internal) return(get_gender_internal(names, prob, threshold))


  ### API data
  # Whole country & unique names
  if(is.null(state) & ln == ln_un){

    # Return
    out <- sapply(1:ln, function(i) get_gender_api(names[i], state, prob = prob, threshold = threshold, pause = pause))
    return(out)
  }

  # Whole country & non-unique names
  if(is.null(state)) {

    # Pick unique names
    gender_pred <- sapply(1:ln_un, function(i) get_gender_api(un_names[i], state, prob = prob, threshold = threshold, pause = pause))

    # Join
    names <- dplyr::tibble(names = names)
    un_names <- dplyr::tibble(names = un_names, prob = gender_pred)

    # Return
    out <- dplyr::left_join(names, un_names, by = c("names"))$prob
    return(out)
  }

  # By state & unique names
  if(ln == ln_un){

    # Return
    state <- get_state(state, ln)
    out <- sapply(1:ln, function(i) get_gender_api(names[i], state[i], prob = prob, threshold = threshold, pause = pause))
    return(out)
  }

  # By state & non-unique names
  state <- get_state(state, ln)
  names <- dplyr::tibble(names = names, state = state)
  dis_names <- dplyr::distinct(names)

  dis_names$prob <- sapply(1:length(dis_names$names), function(i) get_gender_api(dis_names$names[i], dis_names$state[i], prob = prob, threshold = threshold, pause = pause))
  out <- dplyr::left_join(names, dis_names, by = c("names", "state"))$prob

  return(out)
}


# Get individual results from API
get_gender_api <- function(name, state, prob, threshold, pause = pause){


  # API endpoint
  ibge <- "https://servicodados.ibge.gov.br/api/v1/censos/nomes/basica"

  # GET
  females <- get_safe(ibge, query = list(nome = name, regiao = state, sexo = "f"))
  if(pause) Sys.sleep(0.3)
  males <- get_safe(ibge, query = list(nome = name, regiao = state, sexo = "m"))

  # Test responses
  if(is.null(females)) stop("IBGE's API is not responding. Try again later.")
  if(is.null(males)) stop("IBGE's API is not responding. Try again later.")
  res <- test_responses(females, males, prob)
  if(!is.null(res)) return(res)

  # Parse freq
  females <- httr::content(females, as = "parsed")[[1]]$freq
  males <- httr::content(males, as = "parsed")[[1]]$freq

  # Return
  fprob <- females / sum(females, males)
  if(prob) return(fprob)
  round_guess(fprob, threshold)
}



# Get results from internal data
get_gender_internal <- function(names, prob, threshold){

  # Join data
  nms <- dplyr::tibble(nome = names) %>%
    dplyr::left_join(nomes, by = "nome")

  # Return
  if(prob) return(nms$prob_fem)
  round_guess(nms$prob_fem, threshold)
}

